Add IQS039 rule (#2415)
home-assistant/developers.home-assistant: docs/core/integration-quality-scale/rules/iqs039.md

---

> title: "IQS039 - issues and repairs are used effectively"

Reviewer: ```suggestion
title: "IQS039 - repair issues and repair flows are used when user intervention is needed"
```

---

IQS039 is as follows:

issues and repairs are used effectively

## Reasoning

> Issues and repairs are a very user-friendly manner to let the user know something is wrong and that they can do something about it.

Reviewer: Please rename them to "repair issues" and "repair flows".

Issues are just a way to let the user know that they can fix it themselves, while repairs can fix it for them.

Issues and repairs should be actionable and informative about the problem.

> Thus, we should not raise issues for just letting know something is wrong, which they can't fix themselves.

Reviewer: ```suggestion
Thus, we should not raise issues for just letting users know that something is wrong, which they can't fix themselves.
```

## Example implementation

In the example below we have an integration for a locally hosted service.
On boot, we check if we support the version of the service that is running.
If we do not, we raise an issue where we let the user know that they should update their service before they can use the integration again.

`__init__.py`
```python
async def async_setup_entry(hass: HomeAssistant, entry: MyConfigEntry) -> None:
    """Set up the integration from a config entry."""
    client = MyClient(entry.data[CONF_HOST])
    version = await client.get_version()
    if version < MINIMUM_VERSION:
        ir.async_create_issue(
            hass,
            DOMAIN,
            "outdated_version",
            is_fixable=False,
            issue_domain=DOMAIN,
            severity=ir.IssueSeverity.ERROR,
            translation_key="outdated_version",
        )

>         return False

Reviewer: ```suggestion
        raise ConfigEntryError(
            "Version of MyService is %s, which is lower than minimum version %s",
            version,
            MINIMUM_VERSION,
        )
```
